test/rulesets/Base/base.jl

```diff
@@ -179,4 +179,64 @@
         frule_test(clamp, (x, Δx), (y, Δy), (z, Δz))
         rrule_test(clamp, Δk, (x, x̄), (y, ȳ), (z, z̄))
     end
+
+    @testset "non-commutative number (quaternion)" begin
+        function FiniteDifferences.to_vec(q::Quaternion)
+            function Quaternion_from_vec(q_vec)
+                return Quaternion(q_vec[1], q_vec[2], q_vec[3], q_vec[4])
+            end
+            return [q.s, q.v1, q.v2, q.v3], Quaternion_from_vec
+        end
```

Comment:
We should just have this defined in ChainRulesTestUtils.jl
Whole point of that package is to avoid defining re-usable stuff inside the tests.

Its still type piracy there but it makes sense for us to define proper testing functionality for testing this using ChainRulesTestUtils.
https://github.com/JuliaDiff/ChainRulesTestUtils.jl/pull/61

Or we could move it to FiniteDifferences.jl that would also be acceptable, and not type-piracy.